File: Python/boj/Silver/2805.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N, M = map(int, input().split(" "))
    tree_height = sorted(list(map(int, input().split(" "))))
    result = 0

    start = 1
    end = tree_height[N - 1]

    while start <= end:
        sum = 0
        mid = (start + end) // 2

        for h in tree_height:
            if h - mid > 0:
                sum += h - mid

        if sum >= M:
            result = max(result, mid)
            start = mid + 1
        else:
            end = mid - 1

    print(result)

File: JavaScript/boj/Silver/2805.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const [N, M] = input[0].split(' ').map(Number);
const trees = input[1].split(' ').map(Number);
let max = 0;
let left = 1;
let right = Math.max(...trees);

while (left <= right) {
  let mid = Math.floor((left + right) / 2);
  let sum = 0;

  for (let i = 0; i < N; i++) {
    if (trees[i] - mid > 0) sum += trees[i] - mid;
  }

  if (sum >= M) {
    left = mid + 1;
    max = Math.max(max, mid);
  } else right = mid - 1;
}

console.log(max);
